maven-core/src/main/java/org/apache/maven/toolchain/building/DefaultToolchainsBuilder.java

```diff
@@ -66,16 +75,86 @@ public ToolchainsBuildingResult build( ToolchainsBuildingRequest request )
         toolchainsMerger.merge( userToolchains, globalToolchains, TrackableBase.GLOBAL_LEVEL );
         
         problems.setSource( "" );
-        
+
+        userToolchains = interpolate( userToolchains, problems );
+
         if ( hasErrors( problems.getProblems() ) )
         {
             throw new ToolchainsBuildingException( problems.getProblems() );
         }
-        
-        
+
+
         return new DefaultToolchainsBuildingResult( userToolchains, problems.getProblems() );
     }
 
+    private PersistedToolchains interpolate( PersistedToolchains toolchains, ProblemCollector problems )
+    {
+
+        StringWriter stringWriter = new StringWriter( 1024 * 4 );
+        try
+        {
+            toolchainsWriter.write( stringWriter, null, toolchains );
+        }
+        catch ( IOException e )
+        {
+            throw new IllegalStateException( "Failed to serialize toolchains to memory", e );
+        }
+
+        String serializedToolchains = stringWriter.toString();
+
+        RegexBasedInterpolator interpolator = new RegexBasedInterpolator();
+
+        try
+        {
+            interpolator.addValueSource( new EnvarBasedValueSource() );
+        }
+        catch ( IOException e )
+        {
+            problems.add( Problem.Severity.WARNING, "Failed to use environment variables for interpolation: "
+                    + e.getMessage(), -1, -1, e );
+        }
+
+        interpolator.addPostProcessor( new InterpolationPostProcessor()
+        {
+            @Override
+            public Object execute( String expression, Object value )
+            {
+                if ( value != null )
+                {
+                    // we're going to parse this back in as XML so we need to escape XML markup
+                    value = value.toString().replace( "&", "&amp;" ).replace( "<", "&lt;" ).replace( ">", "&gt;" );
```

Comment:
@rscholte
Do we have any library for doing this xml escape ?

Reply:
This line of code is pretty straight forward, at least not worth introducing a new library. Maybe one of the existing libraries has one.
Actually, looking at it I wonder the performance penalty here since we loop over the whole file 3 times. But that kind of optimization deserves a new issue.

Reply:
@rfscholte
Sure. In fact I think it is not worth to add a library for this.
If I understand correctly this code is meant to run only really few times, so there is no need for aggressive optimizations

Reply:
This logic is taken from the DefaultSettingsBuilder, if it already exists in one of the libraries I can change both places.